torch/testing/_internal/composite_compliance.py

```diff
@@ -0,0 +1,174 @@
+        raise RuntimeError(
+            "This operator is not CompositeImplicitAutograd compliant: the "
+            "dtype of the tensor was modified directly without "
+            "going through the PyTorch dispatcher.")
+
+# This is a bit fragile because this needs to be updated whenever a new view op
+# is added. We probably want some sort of introspection to tell that if a
+# torch.* op is a view or not.
+def is_view_fn(func):
+    return func.__name__ in {
+        'as_strided',
+        'detach',
+        'diagonal',
+        'expand',
+        'expand_as',
+        'movedim',
+        'narrow',
+        'permute',
+        'select',
+        'squeeze',
+        'transpose',
+        't',
+        'real',
+        'imag',
+        'view_as_real',
+        'view_as_complex',
+        'unflatten',
+        'unfold',
+        'unsqueeze',
+        'view',
+        'view_as',
+        'unbind',
+        'split',
+        'split_with_sizes',
+        'vsplit',
+        'hsplit',
+        'tensor_split',
+        'chunk',
+        'swapaxes',
+        'slice',
+        '_reshape_alias',
+        '_unsafe_view',
+        '_conj',
+    }
+
+def is_inplace_view_fn(func):
+    return func.__name__ in {
+        'squeeze_',
+        'unsqueeze_',
+        'transpose_',
+        't_',
+    }
+
+class CompositeCompliantTensor(torch.Tensor):
+    elem: torch.Tensor
+
+    __slots__ = ['elem']
+
+    @staticmethod
+    def __new__(cls, elem, *args, **kwargs):
+        # The storage of CompositeCompliantTensor should never be used directly
+        # by a CompositeImplicitAutograd operation; if the CompositeImplicitAutograd
+        # operator attempts to read from the storage without dispatching then it'll
+        # raise a RuntimeError due to it being a meta storage.
+        r = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
+            cls, elem.size(),
+            dtype=elem.dtype, layout=elem.layout,
+            device=elem.device, requires_grad=elem.requires_grad)
+        r.elem = elem
+        return r
+
+    def __repr__(self):
+        return f"CompositeCompliantTensor({self.elem})"
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        def unwrap(e):
+            return e.elem if isinstance(e, CompositeCompliantTensor) else e
+
+        def wrap(e):
+            return CompositeCompliantTensor(e) if isinstance(e, torch.Tensor) else e
+
+        if func.__name__ in ('set_', 'resize_'):
+            raise RuntimeError(
+                f"{func.__name__} is not allowed to be called inside of "
+                f"CompositeImplicitAutograd operators.")
+
+        with no_dispatch():
+            unwrapped_args = tree_map(unwrap, args)
+            unwrapped_kwargs = tree_map(unwrap, kwargs)
+            unwrapped_rs = func(*unwrapped_args, **unwrapped_kwargs)
+            rs = tree_map(wrap, unwrapped_rs)
+
+        if is_view_fn(func):
```

Comment:
Is there a way to check if there's any storage aliasing going on, without having to check if the op is a view op?

E.g. if you have these for pieces of data `args`, `unwrapped_args`, `unwrapped_results`, `results`, then maybe you could do something like:
```
# ... first ensure that args[0] and result[0] are both valid tensors, since all view ops have a first arg with tensor type
if unwrapped_args[0].storage().is_alias_of(unwrapped_results[0].storage()):
```

I'm not 100% that all of those API's are exposed in python though

Reply:
That's a good point, thanks for the suggestion. There are some APIs that are exposed; e.g. we can do
```
unwrapped_results[0]._base is unwrapped_args[0]
```
Let me go play around with this

Reply:
Hmm, I was unable to get the following to work:
- the `unwrapped_results[0]._base is unwrapped_args[0]` doesn't work because torch_dispatch runs after autograd, so the _base field never gets populated (that happens somehow during the autograd logic?)
- binding `is_alias_of` into Python doesn't work, because more than just view functions return aliases (in-place operations is one example). The later code segfaults if presented with an in-place operation, which is not good.

So the conclusion here is: we probably want some introspection at the python level to be able to tell if something is a view operation?

Reply:
> that happens somehow during the autograd logic?

Yes this is part of the InplaceOrView autograd key.

> Is there a way to check if there's any storage aliasing going on, without having to check if the op is a view op?

We actually don't (should not) have multiple storages that look at the same memory. And if the input and output share their storage, they have to alias (we don't have out= variants here). So something like `unwrapped_args[0].storage().data_ptr() == unwrapped_results[0].data_ptr()` would be enough no?

Also because data_ptr() is bad, we could fix the storage python binding to have `unwrapped_args[0].storage() is unwrapped_results[0].storage()` work :D

Reply:
> unwrapped_args[0].storage().data_ptr() == unwrapped_results[0].data_ptr() would be enough no?

It turns out this doesn't work for tensors with 0-size dimensions. The data_ptr() returns 0

> Also because data_ptr() is bad, we could fix the storage python binding to have unwrapped_args[0].storage() is unwrapped_results[0].storage() work :D

Yes, that would be awesome (going to file an issue...)